Improve ImmutableArray.RemoveRange

lateapexearlyspeed opened the pull request:
Fixed #65069 
Will mark as ready when it is done.

lateapexearlyspeed committed:
Initial commit: empty MultiSet.

lateapexearlyspeed committed:
Implement MultiSet.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1399,96 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+        private struct MultiSet
+        {
+            private readonly IEqualityComparer<T> _equalityComparer;
+            private readonly Dictionary<int, LinkedList<(T value, int count)>> _dictionary = new();
```
Since you're encoding a hashtable implementation here, wouldn't it make sense to use an array of buckets directly to avoid all the linked list allocations?

lateapexearlyspeed replied:
Got it, fixed to array-based implementation. 
I referred the basic concept ([bucket headers array and buckets array](https://github.com/dotnet/runtime/blob/c1ee3cdbd123777384c0b9c4ffd704f691508605/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L40)) of generic `Dictionary` and `HashSet` and simplify something based on `ImmutableArray.RemoveRange` scenario, for example: no need to reuse free holes because `Remove()` will only happen after all `Add()` are called. 
Please correct my understanding.

lateapexearlyspeed committed:
Array-based multiset implementation with assumptions:

Call .TryRemove() after all Add() calls are called and not remove item entries from array, so not optimize free-holes from Remove
immutableArrray.RemoveRange should know item count so assume it is not possible to exceed initial capacity (no Resize)

lateapexearlyspeed committed:
Array-based multiset implementation: support extending size.

lateapexearlyspeed committed:
Use MultiSet in ImmutableArray.RemoveRange overloads; Add tests.

lateapexearlyspeed committed:
Remove ValueTuple package.

lateapexearlyspeed commented:
@eiriktsarpalis it is ready for review now, please correct my understand of `MultiSet` implementation for `ImmutableArray.RemoveRange`.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -575,17 +575,8 @@ public ImmutableArray<T> RemoveRange(IEnumerable<T> items, IEqualityComparer<T>?
-            }
-
-            return self.RemoveAtRange(indicesToRemove);
+            ICollection<int>? indicesToRemove = items.TryGetCount(out _) ? self.FindOrderedIndicesToRemoveByMultiSet(items, equalityComparer) : self.FindOrderedIndicesToRemove(items, equalityComparer);
```
I'm not sure I understand why we need to fall back to the old algorithm if we can't determine the size?

lateapexearlyspeed replied:
See previous [comment(61196)](https://github.com/dotnet/runtime/pull/61196#discussion_r793513803)
Here we use `TryGetCount` to test type.

eiriktsarpalis replied:
Maybe add a comment explaining this? What about collections with very large counts?

lateapexearlyspeed replied:
> Maybe add a comment explaining this?

Added.

> What about collections with very large counts?

Not sure, I just feel the fall back code is for true enumerable data which sounds like long `streaming` concept. But if it is already a large reality collection data, that will mean that it was successful to allocate large memory (so same order of magnitude). Please give suggestion.

eiriktsarpalis replied:
One possible approach might be to fall back to the slow implementation once the multiset exceeds a certain capacity dynamically:
```C#
        private ICollection<int>? FindOrderedIndicesToRemoveByMultiSet(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
        {
            var multiSet = new MultiSet(equalityComparer);
            int i = 0;
            foreach (T item in items)
            {
                multiSet.Add(item);
                if (++i == 100) // threshold way above what I would expect is the typical input enumerable.
                {
                      return FindOrderedIndicesToRemove(items, equalityComparer);
                }
            }

            return GetOrderedIndicesToRemoveFor(multiSet);
        }
```
I would strongly encourage you to make any decision based on benchmark numbers, however.

lateapexearlyspeed replied:
// Not check size
           public ICollection<int>? Fast_NoLimit_FindOrderedIndicesToRemoveByMultiSet(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                var multiSet = new MultiSet(equalityComparer);
                foreach (T item in items)
                {
                    multiSet.Add(item);
                }

                return GetOrderedIndicesToRemoveFor(multiSet);
            }

          // Firstly try to get collection count property directly then fall back to iterate to trigger size limit dynamically
           public ICollection<int>? Fast_TryGetCountThenIterateCheckCount(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                if (items.TryGetCount(out int count))
                {
                    if (count <= ThrottleValue)
                    {
                        return Fast_NoLimit_FindOrderedIndicesToRemoveByMultiSet(items, equalityComparer);
                    }

                    return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                }

                return Fast_IterateCheckCount(items, equalityComparer);
            }

          // Firstly try to get collection count property directly then fall back to iterate to trigger size limit compared with actual dictionary.count (dedup same value purpose) dynamically
            public ICollection<int>? Fast_TryGetCountThenIterateCheckCountByDicLength(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                if (items.TryGetCount(out int count))
                {
                    if (count <= ThrottleValue)
                    {
                        return Fast_NoLimit_FindOrderedIndicesToRemoveByMultiSet(items, equalityComparer);
                    }

                    return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                }

                return Fast_IterateCheckCountByDicLength(items, equalityComparer);
            }

          // Iterate to trigger size limit compared with actual dictionary.count (dedup same value purpose) dynamically
            public ICollection<int>? Fast_IterateCheckCountByDicLength(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                var multiSet = new MultiSet(equalityComparer);
                foreach (T item in items)
                {
                    if (multiSet.ItemCount > ThrottleValue)
                    {
                        return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                    }
                    multiSet.Add(item);
                }

                return GetOrderedIndicesToRemoveFor(multiSet);
            }

          // Iterate to trigger size limit dynamically
            public ICollection<int>? Fast_IterateCheckCount(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                var multiSet = new MultiSet(equalityComparer);
                int i = 0;
                foreach (T item in items)
                {
                    if (++i > ThrottleValue)
                    {
                        return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                    }
                    multiSet.Add(item);
                }

                return GetOrderedIndicesToRemoveFor(multiSet);
            }

           private readonly struct MultiSet
           {
                public int ItemCount => _dictionary.Count;
           }

Benchmark condition:
ImmutableArray length = 100, Items length = 100

1. `Items `is true enumerable
![image](https://user-images.githubusercontent.com/72254037/156095858-64f2c2f7-371a-47db-a5d5-9563754d2e6b.png)
2. `Items` is array instance
![image](https://user-images.githubusercontent.com/72254037/156095934-dc08847a-8817-469a-8c2a-de6df991f9ee.png)

Result:
Pure fast method is ~1x faster than slow method
If not hit `throttleValue`, pure fast is almost same time as all other fast methods with throttling check
If hit `throttleValue` so that fall back to slow path, then all fast methods with throttling check are much slower than original slow method, except "`Fast_TryGetCountThenIterateCheckCount()`" and "`Fast_TryGetCountThenIterateCheckCountByDicLength()`" when `items` is Array instance because they can early find situation to fall back to slow path. However they are still slower than original slow method.

So if we need to have size limit, maybe `Fast_TryGetCountThenIterateCheckCountByDicLength()` is better, but not sure what throttling value is better, LOH triggering size is 85k but need to find out it after Dictionary already scales array size (`EnsureCapacity`). Need guidance for that, thanks.
Otherwise, pure fast method can give shortest time in all cases.

eiriktsarpalis replied:
Thanks, I think `Fast_TryGetCountThenIterateCheckCountByDicLength` is ok, although I would be interested in seeing the numbers for various sizes of input enumerables, particularly for low values like 1, 2 or 10 elements. It might be the case that for low enough values creating a MultiSet actually makes things slower. You can execute benchmarks for multiple sizes using BDN's `ParamsAttribute`.

lateapexearlyspeed replied:
Under condition of 100 length ImmutableArray, fast method is slower until items length up to ~30.
![image](https://user-images.githubusercontent.com/72254037/156381903-e9e19378-0f1a-4255-9061-8a019bb6394d.png)
![image](https://user-images.githubusercontent.com/72254037/156382170-0e0f0b2e-ca60-43e0-9fee-33bbace2705e.png)

eiriktsarpalis replied:
Hmmm, these numbers make me think that this optimization is likely not worth our while. I don't think it's worth optimizing for inputs > 30, it's unlikely that this method would be used with so large inputs. I'd probably just close this PR. Thanks for driving the performance investigation!

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1381,183 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+            return indicesToRemove;
+        }
+
+        private class MultiSet
```
Should this be a separate source file?

lateapexearlyspeed replied:
Separated.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1381,183 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+                }
+
+                _entries[_startOffset].SetInfo(item, bucketHeader - 1, 1);
+                bucketHeader = ++_startOffset; // 1-based bucketHeader
```
Why does it need to be assigned to a local?

lateapexearlyspeed replied:
`bucketHeader` is a `ref` variable to actual bucket header in `_bucketHeaders` for `item`. Here it needs to be updated to a new offset.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1381,183 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+
+            private void EnsureCapacity()
+            {
+                const uint arrayMaxLength = 0X7FFFFFC7;
```
I believe we now have an `Array.MaxLength` property you don't need to hardcode this constant.

lateapexearlyspeed replied:
`Array.MaxLength` is from .net 6, so we cannot just use it here.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1381,183 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+                    ref Entry entry = ref _entries[entryIndex];
+                    if (_equalityComparer.Equals(entry.Key, item))
+                    {
+                        if (entry.Count == 0)
```
Nit: not important in this particular application, but a general-purpose MultiSet should probably also remove the entry when count drops to zero. Consider adding a remark stating that we're not doing this.

lateapexearlyspeed replied:
Added.

eiriktsarpalis reviewed (commented):
Would it be possible to write a few benchmarks capturing performance gains (if any) brought by this change?

@stephentoub do you think this performance improvement justifies the inclusion & maintenance of yet another hashtable implementation? It should be noted that it was added since the regular Dictionary does not support `null` keys which is necessary here.

lateapexearlyspeed committed:
Fix comment: separate MultiSet.cs file; add remark for not removing entry with 0 item count.

stephentoub commented:
> @stephentoub do you think this performance improvement justifies the inclusion & maintenance of yet another hashtable implementation?

If no one (besides me asking whether it was an issue) has complained about the algorithmic complexity involved, probably not.

lateapexearlyspeed commented:
> If no one (besides me asking whether it was an issue) has complained about the algorithmic complexity involved, probably not.

@eiriktsarpalis @stephentoub Then is there any chance to proceed ? Maybe put this `MultiSet` into `ImmutableArray` class to specify it is a particular hashtable implementation for `ImmutableArray`, or use solution of [previous commit](https://github.com/dotnet/runtime/pull/65186/commits/b00ba1c2339a61afcf93ae6131063d5e30844bdc) (reuse `Dictionary`) ?

eiriktsarpalis commented:
@lateapexearlyspeed come to think of it, you might be able to work around the null key restriction using something like the following:
```C#
public readonly struct MaybeNullKeyWrapper<TKey>
{
    public TKey? Key { get; init; }
}

public class MaybeNullKeyComparer<TKey> : IEqualityComparer<MaybeNullKeyWrapper<TKey>>
{
    private readonly IEqualityComparer<TKey> _keyComparer;

    public MaybeNullKeyComparer(IEqualityComparer<TKey> keyComparer)
    {
        _keyComparer = keyComparer ?? EqualityComparer<TKey>.Default;
    }

    public int GetHashCode(MaybeNullKeyWrapper<TKey> wrapper) => _keyComparer.GetHashCode(wrapper);
    public bool Equals(MaybeNullKeyWrapper<TKey> left, MaybeNullKeyWrapper<TKey> right) => _keyComparer.Equals(left.Key, right.Key);
}
```
It should then be possible to use a regular dictionary without issue:
```
var multiset = new Dictionary<MaybeNullKeyWrapper<T>, int>(new MaybeNullKeyComparer());

...
```

lateapexearlyspeed committed:
Revert "Array-based multiset implementation with assumptions:"

This reverts commit 62cb66b7056543d82fb6138e6572685f7a5e113f.

# Conflicts:
#	src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

lateapexearlyspeed committed:
Fix comment: introduce NullableKeyWrapper and back to use Dictionary.

lateapexearlyspeed committed:
Pass T typed 'item' into Dictionary more naturally.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/MultiSet.cs:
```diff
@@ -0,0 +1,84 @@
+
+            public int GetHashCode(NullableKeyWrapper obj)
+            {
+                return obj.Key == null ? 0 : _keyComparer.GetHashCode(obj.Key);
```
The null check would violate custom equality comparers that equate null with non-null keys.

lateapexearlyspeed replied:
Yes, but struggle here is as following, please help decide:

1. See regular [HashSet<T>](https://github.com/dotnet/runtime/blob/c32b1f5c7c476aa7a4391558d88b4adf178f3df7/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L1151), it uses same way to handle `null` (no matter comparer) so it will also hit issue you mentioned unless custom comparer's `GetHashCode(null) == 0` also.
2. See definition of `IEqualityComparer<T>.GetHashCode([DisallowNull] T obj)`: interface declaration not allow null argument and compiler nullable check will also find it; although we know Default compare supports and return 0 at least.

eiriktsarpalis replied:
> interface declaration not allow null argument and compiler nullable check will also find it; although we know Default compare supports and return 0 at least.

I think that's fine, I would probably just add a suppression. A good guide is the current behavior of `ImmutableArray.Index`, which works with `null` elements if no equality comparer is specified and will surface any null exceptions a custom equality comparer might throw (it only exercises the `Equals` method, not `GetHashCode`).

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/MultiSet.cs:
```diff
@@ -0,0 +1,84 @@
+
+        private readonly struct NullableKeyWrapper
+        {
+            public readonly T Key;
```
Should be nullable?

lateapexearlyspeed replied:
Changed.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/MultiSet.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace System.Collections.Immutable
+{
+    internal class MultiSet<T>
```
I think the MultiSet class is trivial enough that it can be moved back into the original source file. Also, it can probably be made a struct now.

lateapexearlyspeed replied:
Fixed.

lateapexearlyspeed committed:
Fix comment: Move back MultiSet; define T? key.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -575,17 +577,9 @@ public ImmutableArray<T> RemoveRange(IEnumerable<T> items, IEqualityComparer<T>?
-            }
-
-            return self.RemoveAtRange(indicesToRemove);
+            // Avoid to allocate for true enumerable
```
```suggestion
            // Avoid building a multiset for enumerables of unknown size.
```

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1384,131 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+                }
+            }
+
+            private class NullableKeyWrapperEqualityComparer : IEqualityComparer<NullableKeyWrapper>
```
An opportunity for potential performance optimization is to do the following:
```C#
            private class NullableKeyWrapperEqualityComparer : IEqualityComparer<NullableKeyWrapper>
            {
                private readonly IEqualityComparer<T>? _keyComparer;

                public NullableKeyWrapperEqualityComparer(IEqualityComparer<T>? keyComparer)
                {
                    _keyComparer = keyComparer;
                }

                public int GetHashCode(NullableKeyWrapper key)
                {
                    return _keyComparer is null ? EqualityComparer<TKey>.Default.GetHashCode(key) : _keyComparer.GetHashCode(obj.Key);
                }

                public bool Equals(NullableKeyWrapper x, NullableKeyWrapper y)
                {
                    return  _keyComparer is null ? EqualityComparer<TKey>.Default.Equals(x.Key, y.Key) : _keyComparer.Equals(x.Key, y.Key);
                }
```

The JIT will optimize inlined calls to `EqualityComparer<TKey>.Default.Equals`/`GetHashCode` so this will save us one nested virtual call in the default case.

lateapexearlyspeed replied:
Fixed.

lateapexearlyspeed committed:
Fix comment: inline EqualityComparer<T>.Default.

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1384,131 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+            {
+#if NET6_0_OR_GREATER
+                ref int count = ref CollectionsMarshal.GetValueRefOrAddDefault(_dictionary, item, out _);
+                count++;
```
Might need to guard against overflow?
```suggestion
                checked { count++; }
```

eiriktsarpalis commented on src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:
```diff
@@ -1398,5 +1384,131 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+                ref int count = ref CollectionsMarshal.GetValueRefOrAddDefault(_dictionary, item, out _);
+                count++;
+#else
+            _dictionary[item] = _dictionary.TryGetValue(item, out int count) ? count + 1 : 1;
```
```suggestion
            _dictionary[item] = _dictionary.TryGetValue(item, out int count) ? checked(count + 1) : 1;
```

lateapexearlyspeed commented:
Close this because the new implementation cannot win old implementation under [low item number case](https://github.com/dotnet/runtime/pull/65186#discussion_r817752564).

eiriktsarpalis commented:
Thanks for driving this investigation @lateapexearlyspeed. It was a good learning exercise!
